res/schema.xml

```diff
@@ -447,4 +447,29 @@ METADATA
       ALTER TABLE cues ADD COLUMN source INTEGER DEFAULT 2 NOT NULL;
     </sql>
   </revision>
+  <revision version="30" min_compatible="3">
+    <description>
+      Convert the PredefinedColor id to the actual RGBA value.
+    </description>
+    <sql>
+      <!-- No Color becomes black-->
+      UPDATE cues SET color=4278190080 WHERE color=0;
```

Comment:
How do you differentiate between actual black and no color? IMHO you should make the color column nullable and map no color to `NULL`.

Reply:
In the new implementation No Color is gone. Hotcues always have a color.
Thus we need to assign a default color to the hotcues that had No Color assigned.

Reply:
We need to keep in mind, that a user comming from a earlyer version of Mixxx may has thousands of with cues all turing to black. This will look ugly and forces him to think about colours even if he does not want to use them.  Using a usable default color for it will work, but this creates the situation that this default colour may already has a meaning for the user. A explicited No Colour solved that.


How about using fully transparent for default?

Reply:
I think we need to have a closer look at it because of the migration path. With a default colour black, all exiting cue points will turn to black. I think that is not a good color for default. 

What is the issue you are trying to solve by removing the No Color?

Reply:
We can think of another color.

In my opinion, the only value that No Color brings is that it allows users to use colors as "categories" of hotcues. I.e. blue is where the vocals start, red is where the drop starts, etc.. Thus No Color would mean that the cue has not been assigned to any category yet.
No Color needs to be displayed somehow, so we mapped it to a default color.


However, with the new design, the user can just add another color to the palette and think about it like "the hotcue has not been assigned to any category yet". Since the default color assigned to a hotcue will be configurable, this works well.

In my opinion, the former is simpler from the point of view of the user, but also in code.

Reply:
> all exiting cue points will turn to black

Only the ones with No Color. If a hotcue had a  color assigned we will respect that. The default mixxx palette is the one we designed as our PredefinedColorsSet

Reply:
> Only the ones with No Color.

Wasn't that the default before Mixxx 2.3? So I think all users upgrading will have the default color assigned to all hotcues.

How about we assign the color for cues without a color based on the hotcue number? So hotcue 1 will be assigned to the first color in the palette, hotcue 2 the second color in the palette, and so on.

Reply:
> How about we assign the color for cues without a color based on the hotcue number?

I'm not sure about this. If we do this, then we are assigning e.g Red color to lots of cues in a way that might not fit what the user wants to use Red for. For example, if I want to use Red for cues where vocals start, the automatic assignment won't respect this.

Please keep in mind that we don't have a batch-editing feature for cue points, so we must minimize the decisions we make for the user, because if we change their cues in a way they don't like, there's no easy way they can revert it. From this point of view, assigning to a single color that is outside the palette is the safest choice IMHO.

Reply:
> Please keep in mind that we don't have a batch-editing feature for cue points

So let's add one now. I think it would be fairly easy to implement. The question is where in the GUI to put this.

Reply:
Let's do it in another PR. It's something new.

Reply:
Even a batch feature need to know which cue is a legacy cue without a user set color.

A no colour color will work for me. We can store it as fully transparent in the database and use it non transparent on the waveforms.
The cue button can stay uncolored as a hint for legacy cue.  

If a user creates the button background color on his custom palette, it is his own fault IMHO.

What do you think?

Reply:
> Even a batch feature need to know which cue is a legacy cue without a user set color

Legacy cues with No Color will be mapped to a specific color. You will be able to use this color to identify such cues. I don't understand what use case is possible with the possibility of No Color that is not covered by this solution.

> We can store it as fully transparent in the database and use it non transparent on the waveforms.

I prefer to treat colors in a simple way, without any special cases or hacks. Let's keep it simple.

Reply:
> Legacy cues with No Color will be mapped to a specific color. You will be able to use this color to identify such cues.

I thought about this solution again. You're right, this is probably the best solution as it requires not that much code and will reduce the maintenance burden later on.

Reply:
OK, the proposed black will serves also all use cases. It is a normally useless color and the controller mapping can replace black with the color of its choice to distinguish it from empty cue buttons.

My point wasn't to keep "No Color" as color, more like a color that indicates that "no meaningful color" is set. This can be any color including black or transparent.

Reply:
Do we have a common sense with using black (or transparent) here?

Reply:
Black works for me. Does it for you?

Reply:
Yes. 
What should a controller do when a cue is back? I think it can use the controllers default color and not switch of the cue button, right? Where should we state that so a mapping contributor find the info? We need to be able to distinguish unset cues from not set cues.

Reply:
Controllers can map black to a dimm color if they can. Otherwise yes, they must choose a default color. Maybe the components.js library can handle this for them?

Reply:
Currently the legacy cues are pure red in the skins. Can we pick a different color or actually black? New set cues are palette red, I had expected to have them black as well. 

This fits somehow, because users don't have to care about cue colors in the past and when they may sitll don't want to care they may continue to produce legacy cues like before. 

We may consider to give a default color (including black) for newly created cues in Decks preferences.

Reply:
If I'm not mistaken, hotcue colors have not been part of any Mixxx release yet, so the number of users with manually-colored cue points should be neglible. I'd say let's set uncolored hotcues either to the first color in the palette (red), or assign a palette color based on the number (like the `AutoHotcueColors` option).

IMHO this is a much better solution than mapping all uncolored hotcue to black. It would either need a lot of manual hacking around the issue to display a replacement color, or - if we handle the black cue like any other cue - the performance pads on controllers would more or less display a random color (since most controllers don't have "black" LEDs I assume).

Reply:
Hmmm, I thought the cue colors were included in the latest release. Then we can just remove this DB migration

Reply:
We can't remove the DB migration because we have users who have been using master with the old schema.

Reply:
DB migrations on master are immutable and must not be reverted! They can only be compensated by a subsequent migration.